src/cpp/session/modules/SessionObjectExplorer.cpp

```diff
@@ -0,0 +1,226 @@
+/*
+ * SessionObjectExplorer.cpp
+ *
+ * Copyright (C) 2009-17 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#define R_INTERNAL_FUNCTIONS
+
+#include "SessionObjectExplorer.hpp"
+
+#include <boost/bind.hpp>
+#include <boost/foreach.hpp>
+
+#include <core/Algorithm.hpp>
+#include <core/Error.hpp>
+#include <core/Exec.hpp>
+
+#include <r/RExec.hpp>
+#include <r/RRoutines.hpp>
+#include <r/RSexp.hpp>
+
+#include <session/SessionModuleContext.hpp>
+
+using namespace rstudio::core;
+
+namespace rstudio {
+namespace session {
+namespace modules {      
+namespace explorer {
+
+namespace {
+
+const char * const kExplorerCacheDir = "explorer-cache";
+
+FilePath explorerCacheDir() 
+{
+   return module_context::sessionScratchPath()
+         .childPath(kExplorerCacheDir);
+}
+
+std::string explorerCacheDirSystem()
+{
+   return string_utils::utf8ToSystem(explorerCacheDir().absolutePath());
+}
+
+void removeOrphanedCacheItems()
+{
+   Error error;
+   
+   // if we don't have a cache, nothing to do
+   if (!explorerCacheDir().exists())
+      return;
+   
+   // list source documents
+   std::vector<boost::shared_ptr<source_database::SourceDocument> > docs;
+   error = source_database::list(&docs);
+   if (error)
+   {
+      LOG_ERROR(error);
+      return;
+   }
+   
+   // list objects in explorer cache
+   std::vector<FilePath> cachedFiles;
+   error = explorerCacheDir().children(&cachedFiles);
+   if (error)
+   {
+      LOG_ERROR(error);
+      return;
+   }
+   
+   // remove any objects for which we don't have an associated
+   // source document available
+   BOOST_FOREACH(const FilePath& cacheFile, cachedFiles)
+   {
+      std::string filename = cacheFile.filename();
+      std::string id = filename.substr(0, filename.length() - ::strlen(".rds"));
+      
+      bool foundId = false;
+      BOOST_FOREACH(
+               boost::shared_ptr<source_database::SourceDocument> pDoc,
+               docs)
+      {
+         if (id == pDoc->getProperty("id"))
+         {
+            foundId = true;
+            break;
+         }
+      }
+      
+      if (!foundId)
+      {
+         error = cacheFile.remove();
+         if (error)
+            LOG_ERROR(error);
+      }
+   }
+}
+
+void onShutdown(bool terminatedNormally)
+{
+   if (!terminatedNormally)
+      return;
+   
+   removeOrphanedCacheItems();
```

Comment:
Nit: It'd be slightly more efficient to remove cache entries at the time the tab is closed (instead of waiting and then sweeping them all up at shutdown). Up to you.